Add DESIGN_MATRIX to keywords docs (#10440)
equinor/ert: docs/ert/reference/configuration/keywords.rst

:ref:`DATA_FILE <data_file>`                                            NO                                                                      Provide an ECLIPSE data file for the problem
:ref:`DATA_KW <data_kw>`                                                NO                                                                      Replace strings in ECLIPSE .DATA files
:ref:`DEFINE <define>`                                                  NO                                                                      Define keywords with config scope

> :ref:`DESIGN_MATRIX <design_matrix>`                                    NO                                                                      Add design matrix parameters as a single group

Reviewer: ```suggestion
:ref:`DESIGN_MATRIX <design_matrix>`                                    NO                                                                      Add design matrix parameters from a spreadsheet
```

DESIGN_MATRIX
-------------
.. _design_matrix:

DESIGN_MATRIX is used to read and validate parameters given in XLSX-format.
:code:`DESIGN_MATRIX` supports 1 positional argument, which points to a XLSL file.

*Example:*

::

        DESIGN_MATRIX poly_design.xlsx

Additionally, there are two optional named arguments:

::

        DESIGN_MATRIX <file> DESIGN_SHEET:<name_of_design_sheet> DEFAULT_SHEET:<name_of_default_sheet>

where:

1. DESIGN_SHEET:<name_of_design_sheet> - the name of the :ref:`design_sheet` (defaults to DesignSheet)

2. DEFAULT_SHEET:<name_of_default_sheet> - the name of the :ref:`default_sheet`, which if not present will not be included

*Example:*

::

        DESIGN_MATRIX poly_design.xlsx DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet

The XLSL file must contain a design sheet, where in the columns represents different parameters and rows represent realizations.

*Example of a design sheet:*

.. _design_sheet:

.. list-table:: design sheet
   :widths: 30 30 30 30
   :header-rows: 1

   * - REAL
     - a
     - b
     - c
   * - 0
     - 1
     - 1
     - 2
   * - 1
     - 1
     - 1
     - 2
   * - 3
     - 1
     - 1
     - 3

> The XLSL file can optionally contain a default sheet, where in there are two columns, the first column specifies parameter names and the second

Reviewer: ```suggestion
The XLSL file can optionally contain a default sheet, where there are two columns, the first column specifies parameter names and the second
```

> their default values distributed across all the realizations defined by :ref:`design_sheet`. In the following example the realization 0,1 and 3 will

Reviewer: ```suggestion
default values distributed across all the realizations defined by :ref:`design_sheet`. In the following example the realization 0,1 and 3 will
```

contain parameters d and e, where d=0 and e=1 for all of the them.

*Example of a default sheet:*

.. _default_sheet:

.. list-table:: default sheet
   :widths: 100 100
   :width: 60%

   * - d
     - 0
   * - e
     - 1

> If the :ref:`design_sheet` contains column `REAL`, ert will automatically set all the run models only for the realizations
> specified in `REAL` column; i.e.; 0,1 and 3 in the example. If the `REAL` column is not present, ert will enumerate individual rows
> as realization; i.e. 0,1 and 2 in the example.

Reviewer: ```suggestion
If the :ref:`design_sheet` contains column `REAL`, ert will automatically set the active realizations to what is
specified in `REAL` column; i.e.; 0,1 and 3 in the example. If the `REAL` column is not present, ert will enumerate individual rows
as realization; i.e. 0,1 and 2 in the example.
```

> :code:`DESIGN_MATRIX` supports multiple instances of the keyword given that
>  - the realizations overlap on each instance of the keyword.
>  - the parameter names are either unique or the values need to be the same for the overlapping parameters in the different instances of the keyword.

Reviewer: ```suggestion
Multiple :code:`DESIGN_MATRIX` keywords can be added to the configuration file, and ert will load of the design sheets, ert will validate that:
 - the realizations overlap on each instance of the keyword.
 - the parameter names are either unique or the values need to be the same for the overlapping parameters in the different instances of the keyword.
```

The combination with :ref:`GEN_KW <gen_kw>` parameters is supported. In case of overlapping names, eg. the ert config would contain:

::

        GEN_KW COEFFS coeff_priors
        DESIGN_MATRIX poly_design.xlsl DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet

wherein coeff_priors

::

        a UNIFORM 0 1
        b UNIFORM 0 2
        c UNIFORM 0 5
        d UNIFORM 0 2
        e UNIFORM 0 5

the :ref:`GEN_KW <gen_kw>` group COEFFS would remain, but the values would be read from the design matrix.

> In case of partial overlap (eg. parameter e would be missing from coeff_priors), this will yield a validation error.

Reviewer: Is this true? Isnt it just a warning?

Author: Yes, this will get a ValidationError:
https://github.com/equinor/ert/blob/3b89529e7f9d00538ced573d684e78adaab768b5/src/ert/config/ert_config.py#L860-L867

Reviewer: Can you provide an example? Because it confuses me a bit, if I have: `COEFFS:A`, `COEFFS:b` and `COEFFS:C`, and only A and B in my design sheet, that will work? But that is not the case we are talking about here?

    See the :ref:`DATA_KW <data_kw>` keyword which can be used to utilize more template
    functionality in the Eclipse/flow datafile.

DESIGN_MATRIX
-------------
.. _design_matrix:

> DESIGN_MATRIX is meant to be used to read, validate and internalize into ert storage a design matrix parameters given in XLSX-format.

Reviewer: ```suggestion
DESIGN_MATRIX is used to read and validate parameters in XLSX-format.
```

:code:`DESING_MATRIX` supports 1 positional argument and up to two named arguments:

1. Positional argument: XLSX file containing design and default sheet with parameters

Optional named arguments:

> 1. The name of the design sheet (defaults to DesignSheet)

Reviewer: Should have the name of the named argument

> 2. The name of the default sheet (defaults to DefaultSheet)

Reviewer: Should have the name of the named argument

*Example:*

::

>         -- DESIGN_MATRIX poly_design.xlsx DESIGN_SHEET:DesignSheet DEFAULT_SHEET:DefaultSheet

Reviewer: Suggest starting with an example with just a positional argument, and then have this example afterwards.

> If the design sheet contains column `REAL`, the active realization mask is set according to the given realizations.

Reviewer: What is a design sheet? How does it look like? What is a default sheet, what does it look like?

Reviewer: Also mark that default sheet is required, though can be empty (that should probably be an issue that we fix)

Author: Fixed
